File: apps/server/src/services/langgraph/toolStreamMapper.ts
import type { BaseMessage } from '@langchain/core/messages';
import type { LangChainStreamEvent } from '../chat/streamEventTypes.js';
import { createTimelineEvent } from '../chat/timelineEvents.js';
import { safeStringify } from '../chat/safeStringify.js';

type MessageTextPart = {
  text?: unknown;
};

// 读取消息类型
const readMessageType = (message: BaseMessage): string | undefined => {
  const maybeGetType = (message as { _getType?: unknown })._getType;
  if (typeof maybeGetType === 'function') {
    return maybeGetType.call(message) as string;
  }

  const maybeType = (message as { type?: unknown }).type;
  return typeof maybeType === 'string' ? maybeType : undefined;
};

// 读取消息文本
export const readMessageText = (message: BaseMessage | undefined): string => {
  if (!message) return '';
  const raw: unknown = message.content;

  if (typeof raw === 'string') return raw;
  if (Array.isArray(raw)) {
    return raw
      .map((item) => {
        if (typeof item === 'string') return item;
        if (typeof item === 'object' && item !== null) {
          const maybeText = (item as MessageTextPart).text;
          return typeof maybeText === 'string' ? maybeText : '';
        }
        return '';
      })
      .join('');
  }

  if (typeof raw === 'object' && raw !== null) {
    const maybeText = (raw as MessageTextPart).text;
    return typeof maybeText === 'string' ? maybeText : '';
  }

  return '';
};

// 读取工具调用
const readToolCalls = (message: BaseMessage) => {
  const maybeToolCalls = (message as { tool_calls?: unknown }).tool_calls;
  if (!Array.isArray(maybeToolCalls)) {
    return [];
  }

  const parsed: Array<{ name: string; args?: unknown; id?: string }> = [];

  for (const item of maybeToolCalls) {
    if (typeof item !== 'object' || item === null) {
      continue;
    }

    const call = item as { name?: unknown; args?: unknown; id?: unknown };
    if (typeof call.name !== 'string') {
      continue;
    }

    parsed.push({
      name: call.name,
      args: call.args,
      id: typeof call.id === 'string' ? call.id : undefined,
    });
  }

  return parsed;
};

// 将 LangGraph 更新映射为时间线事件
export const mapLangGraphUpdateToTimelineEvents = (
  nodeName: string,
  messages: BaseMessage[],
): LangChainStreamEvent[] => {
  const events: LangChainStreamEvent[] = [];

  for (const message of messages) {
    const type = readMessageType(message);

    if (nodeName === 'agent' && (type === 'ai' || type === 'assistant')) {
      const toolCalls = readToolCalls(message);
      for (const call of toolCalls) {
        events.push(
          createTimelineEvent(
            'tool',
            'start',
            `工具 ${call.name} 开始执行。`,
            call.name,
            safeStringify(call.args ?? {}),
          ),
        );
      }
    }

    if (nodeName === 'tools' && type === 'tool') {
      const toolName = (message as { name?: unknown }).name;
      const resolvedName = typeof toolName === 'string' ? toolName : 'unknown_tool';
      const output = readMessageText(message);
      const statusType = (message as { status?: unknown }).status;
      const failed = statusType === 'error';

      events.push(
        createTimelineEvent(
          'tool',
          'end',
          failed ? `工具 ${resolvedName} 执行失败。` : `工具 ${resolvedName} 执行完成。`,
          resolvedName,
          undefined,
          failed ? undefined : output,
          failed ? output || 'tool execution failed' : undefined,
        ),
      );
    }
  }

  return events;
};

const isHumanMessage = (message: BaseMessage): boolean => {
  const type = readMessageType(message);
  return type === 'human' || type === 'user';
};

const isAssistantMessage = (message: BaseMessage): boolean => {
  const type = readMessageType(message);
  return type === 'ai' || type === 'assistant';
};

/** 全量 state 中最后一条助手消息（含历史轮次，勿单独用于流式 delta）。 */
export const findLatestAssistantText = (messages: BaseMessage[]): string => {
  const lastAssistant = [...messages].reverse().find((msg) => isAssistantMessage(msg));
  return readMessageText(lastAssistant);
};

/**
 * 仅取「最后一条用户消息之后」的助手正文，避免 LangGraph checkpoint 把上一轮回复当成本轮增量。
 */
export const findCurrentTurnAssistantText = (messages: BaseMessage[]): string => {
  let lastHumanIndex = -1;

  for (let index = messages.length - 1; index >= 0; index -= 1) {
    if (isHumanMessage(messages[index]!)) {
      lastHumanIndex = index;
      break;
    }
  }

  const afterHuman = messages.slice(lastHumanIndex + 1);
  const turnAssistants = afterHuman.filter((msg) => isAssistantMessage(msg));
  const lastTurnAssistant = turnAssistants[turnAssistants.length - 1];

  return readMessageText(lastTurnAssistant);
};

File: apps/server_py/app/agent/tool_stream_mapper.py
from __future__ import annotations

from typing import Any

from app.agent.messages import read_message_text, safe_stringify
from app.agent.timeline import TimelineEvent, create_timeline_event


def _read_tool_calls(message: Any) -> list[dict[str, Any]]:
    tool_calls = getattr(message, "tool_calls", None)
    if not isinstance(tool_calls, list):
        return []

    parsed: list[dict[str, Any]] = []
    for item in tool_calls:
        if not isinstance(item, dict):
            continue
        name = item.get("name")
        if not isinstance(name, str):
            continue
        parsed.append(
            {
                "name": name,
                "args": item.get("args"),
                "id": item.get("id"),
            }
        )
    return parsed


def map_langgraph_update_to_timeline_events(
    node_name: str,
    messages: list[Any],
) -> list[TimelineEvent]:
    events: list[TimelineEvent] = []

    for message in messages:
        message_type = getattr(message, "type", None)

        if node_name == "agent" and message_type in {"ai", "assistant"}:
            for call in _read_tool_calls(message):
                events.append(
                    create_timeline_event(
                        "tool",
                        "start",
                        f"工具 {call['name']} 开始执行。",
                        tool_name=call["name"],
                        tool_input=safe_stringify(call.get("args") or {}),
                    )
                )

        if node_name == "tools" and message_type == "tool":
            tool_name = getattr(message, "name", None)
            resolved_name = tool_name if isinstance(tool_name, str) else "unknown_tool"
            output = read_message_text(message)
            status = getattr(message, "status", None)
            failed = status == "error"
            events.append(
                create_timeline_event(
                    "tool",
                    "end",
                    (
                        f"工具 {resolved_name} 执行失败。"
                        if failed
                        else f"工具 {resolved_name} 执行完成。"
                    ),
                    tool_name=resolved_name,
                    tool_output=None if failed else output,
                    tool_error=output if failed else None,
                )
            )

    return events
